openeogeotrellis/utils.py

```diff
@@ -802,3 +808,66 @@ def wait_till_path_available(path: Path):
         else:
             logger.warning(f"Path is not available after {max_tries} tries: {path}")
             return  # TODO: Throw error instead?
+
+
+@dataclasses.dataclass(frozen=True)
+class FileChangeFingerprint:
+    """
+    FileChangeFingerprint can help to detect changes in files. This is useful when there is a need to reload a file if
+    its content has changed.
+
+    This is based on wisdom from a blog: https://apenwarr.ca/log/20181113
+    """
+    mtime: float
+    size: int
+    inode_nr: int
+    file_mode: int
+    owner_uid: int
+    owner_gid: int
+
+    @classmethod
+    def from_file(cls, file_path: Path) -> "FileChangeFingerprint":
+        if file_path.exists():
+            fs = os.stat(file_path)
+            return cls(
+                mtime=fs.st_mtime,
+                size=fs.st_size,
+                inode_nr=fs.st_ino,
+                file_mode=fs.st_mode,
+                owner_uid=fs.st_uid,
+                owner_gid=fs.st_gid,
+            )
+        else:
+            # For missing files we take a fingerprint that is impossible for an existing file which can be used as
+            # a sentinel
+            return cls(
+                mtime=0.0,
+                size=0,
+                inode_nr=0,
+                file_mode=0,
+                owner_uid=0,
+                owner_gid=0,
+            )
+
+
+_last_config_reload: dict[Path, FileChangeFingerprint] = {}
+
+
+def get_file_reload_register_func_if_changed(file_path: Path, resolve=True) -> Optional[Callable[[], None]]:
```

Comment:
I also feel there is something off here with this helper, but I can't really put my finger on it.
Maybe it's because it depends on a global `_last_config_reload` so it only works properly when used from some kind of singleton (or single owner) context. Which means it might not belong yet in a generic utils module

Reply:
Is your concern the shared state and how it could give issues if multiple observe the same file. Because that was indeed not tackled in the implementation. I refactored it to be a watcher that can be instantiated. This way it can be safely used by multiple components.

https://github.com/Open-EO/openeo-geopyspark-driver/pull/969/commits/c0d631d691fdd2e139a1887b8a3fc971c1566f93

I can equally well move it to the identity module but it really is utility code and not specific to identity code.